### PR TITLE
Qiskit compatible mapping

### PR DESCRIPTION
**Pull request by S-Linde**

### CODE REVIEW

File: qgym/benchmarks/metrics/routing_metrics.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                initial_interaction_qubits.append(qubit2)
+
+        initial_qubit_fidelities = []
+        for qubit in initial_interaction_qubits:
+            qubit_fidelity = 1
+            for qubit1, qubit2 in initial_interaction_circuit:
+                if qubit == qubit1 or qubit == qubit2:
+                    qubit_fidelity *= connection_graph[qubit1][qubit2]["weight"]
+            initial_qubit_fidelities.append(qubit_fidelity)
+        initial_qubit_fidelity_mean = mean(initial_qubit_fidelities)
```
Comment (rturrado) [inline qgym/benchmarks/metrics/routing_metrics.py:195]: All this code is basically a _transform_ of `initial_interaction_qubits` into `initial_qubit_fidelities`, i.e., we walk the former and tranform it into the latter.

And the transformation function  is a _reduction_, i.e., we walk `initial_interaction_circuit` and come up with a fidelity value.

```
initial_qubit_fidelities = [
    functools.reduce(
        lambda fidelity, tuple : fidelity * connection_graph[ t[0] ][ t[1] ][ "weight" ] if q in tuple else fidelity,
        initial_interaction_circuit,
        1
    )
    for q in initial_interaction_qubits
]
```

Transformations can be implemented as list comprehensions in Python.

For reductions, we have `functools.reduce`.

In general, thinking in terms of algorithms instead of for loops:
1. Is more elegant, more concise.
2. Most importantly, it expresses intent better.

File: qgym/benchmarks/metrics/routing_metrics.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+        final_qubit_fidelities = []
+        for qubit in initial_interaction_qubits:
+            qubit_fidelity = 1
+            for qubit1, qubit2 in final_interaction_circuit:
+                if qubit == qubit1 or qubit == qubit2:
+                    qubit_fidelity *= connection_graph[qubit1][qubit2]["weight"]
+            final_qubit_fidelities.append(qubit_fidelity)
```
Comment (rturrado) [inline qgym/benchmarks/metrics/routing_metrics.py:203]: Same as above.

Even the code can be reduce to a one-liner, I would probably put it into a meaningful function, e.g., `calculate_fidelities`, so this function can be reduced to a few calls, like 5 or 6 lines of code.

File: qgym/envs/initial_mapping/initial_mapping_visualiser.py [hunk 3]
```diff
@@ -258,7 +258,8 @@ def _draw_mapped_graph(self, screen: Surface, mapped_graph: nx.Graph) -> None:
             elif mapped_graph.edges[u, v]["color"] == "gray":
                 color = self.colors["unused_edge"]
             else:
-                raise ValueError("Unknow color")
+                msg = "unknown color"
```
Comment (rturrado) [inline qgym/envs/initial_mapping/initial_mapping_visualiser.py:261]: Yes! haha

File: qgym/benchmarks/metrics/initial_mapping_metrics.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            return cast(NDArray[np.float64], np.array(results, dtype=np.float64))
+        if return_type == "quartiles":
+            return cast(
+                NDArray[np.float64],
+                np.quantile(results, [0, 0.25, 0.5, 0.75, 1], axis=1),
+            )
+        if return_type == "median":
+            return cast(NDArray[np.float64], np.median(results, axis=1))
+        if return_type == "mean":
+            return cast(NDArray[np.float64], np.mean(results, axis=1))
```
Comment (rturrado) [inline qgym/benchmarks/metrics/initial_mapping_metrics.py:142]: What about having a ReturnType, or RunResult, or BenchmarkResult:

- which holds internally this `NDArray[np.float64]`, and
- offers a series of `as_raw`, `as_quartiles`, `as_median`, `as_mean` API?

So the caller wouldn't do `run(..., "quartiles")`, but `run(...).as_quartiles()`.

Let's say this code shouldn't belong here, even if it's the only point in the code where it is used:

- `run` would receive a fewer parameter, wouldn't check that parameter, and would return a result.
- For  the caller the code wouldn't be much more complicated. They would understand that `run` just returns a result and that the Result type offers different casts.
- Arguably, the casts wouldn't be either part of the Result type, they should just be part of the language. But they are complex lines, very much focus in the Result type, and we can consider them helpers.

File: qgym/benchmarks/metrics/initial_mapping_metrics.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        """
+
+
+class InitialMappingBenchmarker:
+
+    def __init__(
+        self,
+        generator: GraphGenerator | None = None,
+        *,
+        metrics: Iterable[InitialMappingMetric] | InitialMappingMetric,
```
Comment (rturrado) [inline qgym/benchmarks/metrics/initial_mapping_metrics.py:86]: Could this always be an Iterable? We would force the caller to put a single metric into a list, but hey, that's our API. It would simplify the code internally.

File: qgym/benchmarks/metrics/initial_mapping_metrics.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    ) -> None:
+        """Init of the :class:`InitialMappingBenchmarker` class.
+
+        Args:
+            generator: Interaction graph generator to use during benchmarking
+            metrics: Metrics to compute.
+        """
+        self.generator = BasicGraphGenerator() if generator is None else generator
+        self.metrics = (
+            (metrics,) if isinstance(metrics, InitialMappingMetric) else tuple(metrics)
```
Comment (rturrado) [inline qgym/benchmarks/metrics/initial_mapping_metrics.py:96]: Basically, we wouldn't do this ourselves. It should be done on the caller side.

File: qgym/benchmarks/metrics/initial_mapping_metrics.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+@runtime_checkable
+class InitialMappingMetric(Protocol):
+    def compute(self, interaction_graph: nx.Graph, mapping: ArrayLike) -> float: ...
+
+
+@runtime_checkable
+class Mapper(Protocol):
+    def compute_mapping(
+        self, circuit: QuantumCircuit | DAGCircuit
```
Comment (rturrado) [inline qgym/benchmarks/metrics/initial_mapping_metrics.py:35]: This looks to me like a code smell. Can't we have a more generic entity that works as an umbrella for these two types of circuits?

Generic code should lead to simpler code.

Code like this should lead to if/elses, checks in the implementation.

**Review by rturrado — COMMENTED** [review]
I've started with some comments, more tomorrow!

I still have no idea what all this code does, but it looks quite good overall!

File: qgym/generators/qiskit_circuit.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        n_layers: SupportsInt = 1,
+        seed: Generator | SupportsInt | None = None,
+    ) -> None:
+        """Init of the :class:`MaxCutQAOAGenerator`.
+
+        The QAOA circuit is made for the MaxCut problem defined on a randomly generated
+        Erdős-Rényi graph of size `n_nodes` with and edge probability of
+        `edge_probability`. The QAOA circuit is a simple circuit which initializes all
+        qubits in the |+⟩ state. Next repeated cost and mixer layers are added
+        `n_layers` times. The mixer layer consists of parameterized Rx gate on all
```
Comment (rturrado) [inline qgym/generators/qiskit_circuit.py:32]:
```suggestion
        `n_layers` times. The mixer layer consists of parameterized Rx gates on all
```

File: qgym/utils/qiskit_utils.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            :class:`~qiskit.dagcircuit.DAGCircuit` representations are accepted.
+
+    Returns:
+        Interaction graph of the `circuit`.
+    """
+    dag = _parse_circuit(circuit)
+
+    if dag.multi_qubit_ops():
+        msg = "no 3+ qubit operations are supported"
+        raise ValueError(msg)
```
Comment (rturrado) [inline qgym/utils/qiskit_utils.py:31]: I would write this in a single line.
```suggestion
        raise ValueError("no 3+ qubit operations are supported")
```

File: qgym/utils/qiskit_utils.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+    qreg_to_int = _get_qreg_to_int_mapping(dag)
+    interaction_graph: nx.Graph = nx.empty_graph(dag.num_qubits())
+
+    for op_node in dag.op_nodes(include_directives=False):
+        if len(op_node.qargs) == 1:
+            continue
+        qubit1 = qreg_to_int[op_node.qargs[0]]
+        qubit2 = qreg_to_int[op_node.qargs[1]]
+        interaction_graph.add_edge(qubit1, qubit2)
```
Comment (rturrado) [inline qgym/utils/qiskit_utils.py:41]: This can probably be writen as:

```
interaction_graph.add_nodes_from([
    ( qreg_to_int[op_node.qargs[0]], qreg_to_int[op_node.qargs[1]] )
    for op_node in dag.op_nodes(include_directives=False) if len(op_node.qargs) == 1
])
```

File: qgym/utils/qiskit_utils.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    is compatible with the qgym Mapper protocol, which is required for the qgym
+    benchmarking tools.
+    """
+
+    def __init__(self, qiskit_mapper: Layout) -> None:
+        """Init of the :class:`QiskitMapperWrapper`.
+
+        Args:
+            qiskit_mapper: The qiskit mapper (:class:`~qiskit.transpiler.Layout`) to
+                wrap.
```
Comment (rturrado) [inline qgym/utils/qiskit_utils.py:70]: I would consider increasing the line width wrap limit from 90 to 120.

File: tests/benchmarks/metrics/test_routing_metrics.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+def test_distance_ratio_loss() -> None:
+    quality_metric = RoutingSolutionQuality(circuit=circuit_1)
+    # TODO
+    initial_number_of_gates = circuit_1.count_ops()
+    swaps_added = 2
+    loss = quality_metric.gates_ratio_loss(swaps_added=swaps_added)
+
+    assert loss == (initial_number_of_gates + swaps_added) / initial_number_of_gates
+"""
```
Comment (rturrado) [inline tests/benchmarks/metrics/test_routing_metrics.py:39]: Why is this commented out?

Comment (rturrado) [reply tests/benchmarks/metrics/test_routing_metrics.py:39]: Maybe worth creating an issue with the TODO, and remove this uncommented code (or whole file altogether)?

**Review by rturrado — COMMENTED** [review]
I think I'm done now.

Brilliant work!

File: qgym/utils/qiskit_utils.py (hunk 9 above)
Comment (S-Linde) [reply qgym/utils/qiskit_utils.py:31]: While researching `ruff` for `OpenSquirrel` one of the linter options specifically split this into two lines to produce nicer error messages ([see](https://docs.astral.sh/ruff/rules/raw-string-in-exception/)) and now I think this is a bit nicer for the user.

File: qgym/utils/qiskit_utils.py (hunk 10 above)
Comment (S-Linde) [reply qgym/utils/qiskit_utils.py:41]: Yes, it's rewritten to use the comprehension

File: qgym/utils/qiskit_utils.py (hunk 11 above)
Comment (S-Linde) [reply qgym/utils/qiskit_utils.py:70]: Currently we use `black` as our formatter and as they say right on the front page:
[Pro-tip: If you're asking yourself "Do I need to configure anything?" the answer is "No". Black is all about sensible defaults. Applying those defaults will have your code in compliance with many other Black formatted projects.](https://pypi.org/project/black/#:~:text=Pro%2Dtip%3A%20If%20you%27re%20asking%20yourself%20%22Do%20I%20need%20to%20configure%20anything%3F%22%20the%20answer%20is%20%22No%22.%20Black%20is%20all%20about%20sensible%20defaults.%20Applying%20those%20defaults%20will%20have%20your%20code%20in%20compliance%20with%20many%20other%20Black%20formatted%20projects.)

If we at some point move away from `black`, we might change this.

File: tests/benchmarks/metrics/test_routing_metrics.py (hunk 12 above)
Comment (S-Linde) [reply tests/benchmarks/metrics/test_routing_metrics.py:39]: This is code that Willem will work on in the future. Currently, it doesn't hurt to have it on develop, but I 100% agree that it should be removed before merging with `main`.

File: qgym/utils/qiskit_utils.py (hunk 9 above)
Comment (rturrado) [reply qgym/utils/qiskit_utils.py:31]: Fair enough. Thanks for the explanation!

File: tests/benchmarks/metrics/test_routing_metrics.py (hunk 12 above)
Comment (rturrado) [reply tests/benchmarks/metrics/test_routing_metrics.py:39]: What about 1) removing it and 2) creating an issue saying, "add this piece of code when needed in the future..."?

File: qgym/utils/qiskit_utils.py (hunk 11 above)
Comment (rturrado) [reply qgym/utils/qiskit_utils.py:69]: Can we tell `black` to use a line width of 120, i.e., is line width configurable within `black`?

File: qgym/utils/qiskit_utils.py (hunk 10 above)
Comment (rturrado) [reply qgym/utils/qiskit_utils.py]: Rewritten, and properly, using `add_edges_from` and not `add_nodes_from`. Thanks!

File: qgym/benchmarks/metrics/initial_mapping_metrics.py (hunk 7 above)
Comment (rturrado) [reply qgym/benchmarks/metrics/initial_mapping_metrics.py]: You explained this to me already. `QuantumCircuit` and `DAGCircuit` are not classes of our own (they come from external libraries such as Qiskit), so we cannot really have them subclass a generic class of ours.
